compiler/ilgen/VirtualMachineState.hpp

```diff
@@ -0,0 +1,123 @@
+ * code.
+ *
+ * MakeCopy() is typically called when control flow edges are created: the current vm state
+ * must also become the initial vm state at the target of the control flow edge and be
+ * able to evolve independently from the current vm state.
+ *
+ * MergeInto() is typically called when one compiled code path needs to merge into a second
+ * compiled code path (think the bottom of an if-then-else control flow diamond). At the
+ * merge point, the locations of all aspects of the simulated machine state must be
+ * identical so that the code below the merge point will compute correct results.
+ * MergeInto() typically causes the current simulated machine state to be written to the
+ * locations that were used by an existing simulated machine state. The "other" object
+ * passed to MergeInto() must have the exact same shape (number of components and each
+ * component must also match in its type).
+ *
+ * Language compilers should extend this base class, add instance variables for all
+ * necessary virtual machine state variables (using classes like VirtualMachineRegister
+ * and VirtualMachineOperandStack) and then implement Commit(), Reload(), MakeCopy(),
+ * and MergeInto() to call Commit(), Reload(), MakeCopy(), and MergeInto(), respectively, on
+ * each individual instance variable. It feels like boilerplate, but this approach makes
+ * it easy for the compiler to reference individual virtual machine state variables and
+ * at the same time permits complete flexibility in how the VM-wide Commit(), Reload(),
+ * MakeCopy(), and MergeInto() operations can be implemented.
+ */
+
+class VirtualMachineState
+   {
+   public:
+
+   /**
+    * @brief Cause all simulated aspects of the virtual machine state to become real.
+    * @param b builder object where the operations will be added to change the virtual machine state.
+    *
+    * The builder object b is assumed to be along a control flow path transitioning
+    * from compiled code to the interpreter. Base implementation does nothing.
+    */
+   virtual void Commit(TR::IlBuilder *b) { }
+
+   /**
+    * @brief Load the current virtual machine state into the simulated variables used by compiled code.
+    * @param b builder object where the operations will be added to reload the virtual machine state.
+    *
+    * The builder object b is assumed to be along a control flow path transitioning
+    * from the interpreter to compiled code. Base implementation does nothing.
+    */
+   virtual void Reload(TR::IlBuilder *b) { }
+
+   /**
+    * @brief create an identical copy of the current object.
+    * @returns the copy of the current object
+    *
+    * Typically used when propagating the current state along a flow edge to another builder to
+    * capture the input state for that other builder.
+    * Default implementation simply returns the current object.
+    */
+   virtual VirtualMachineState *MakeCopy() { return this; }
```

Comment:
Maybe, but since VirtualMachineState is just an interface, it's unlikely this would be the right answer for any subclass and would just have to be reimplemented anyway. At least "this" is guaranteed to be the expected kind of object for any subclass that inherits this implementation, and VirtualMachineRegister and VirtualMachineRegisterInStruct (currently) can reuse it.